### DISCUSSION TITLE
" ChannelParams_t Channels[]" The value is broken.

### DISCUSSION THREAD

**Discussion by npenny01** [Lora-net/LoRaMac-node#1209] [Q&A]
Environment
--------------------------
- LoRaMacNode v4.4.0
- LoRaWAN specification 1.0.2 and LoRaWAN Regional Parameters v1.0.2rB
- MCU : STM32G473RE
- IDE : keil MDK-ARM v5.27

Error occurrence part
-------------------------------------------

```
// Global attributes
/*!
 * LoRaMAC channels
 */
static ChannelParams_t Channels[KR920_MAX_NB_CHANNELS];
```

```
bool RegionKR920TxConfig( TxConfigParams_t* txConfig, int8_t* txPower, TimerTime_t* txTimeOnAir )
{
    int8_t phyDr = DataratesKR920[txConfig->Datarate];
    int8_t txPowerLimited = LimitTxPower( txConfig->TxPower, Bands[Channels[txConfig->Channel].Band].TxMaxPower, txConfig->Datarate, ChannelsMask );
    uint32_t bandwidth = GetBandwidth( txConfig->Datarate );
    float maxEIRP = GetMaxEIRP( Channels[txConfig->Channel].Frequency );
    int8_t phyTxPower = 0;

        // Take the minimum between the maxEIRP and txConfig->MaxEirp.
        // The value of txConfig->MaxEirp could have changed during runtime, e.g. due to a MAC command.
        maxEIRP = MIN( txConfig->MaxEirp, maxEIRP );

    // Calculate physical TX power
    phyTxPower = RegionCommonComputeTxPower( txPowerLimited, maxEIRP, txConfig->AntennaGain );
			
		if( ( Channels[txConfig->Channel].Frequency != 922100000 ) && ( Channels[txConfig->Channel].Frequency != 922300000 ) && \
				( Channels[txConfig->Channel].Frequency != 922500000 ) && ( Channels[txConfig->Channel].Frequency != 922700000 ) && \
				( Channels[txConfig->Channel].Frequency != 922900000 ) && ( Channels[txConfig->Channel].Frequency != 923100000 ) && \
				( Channels[txConfig->Channel].Frequency != 923300000 ) ) 
		{
				DPRINTF("[error]TxFreq : %d\n", Channels[txConfig->Channel].Frequency );
				DPRINTF("[error]RxFreq : %d\n", Channels[txConfig->Channel].Rx1Frequency );
				DPRINTF("[error]Band : %d\n", Channels[txConfig->Channel].Band );
				DPRINTF("[error]DrRangeMin : %d\n", Channels[txConfig->Channel].DrRange.Fields.Min );
				DPRINTF("[error]DrRangeMax : %d\n", Channels[txConfig->Channel].DrRange.Fields.Max );
			
				DPRINTF("AntennaGain : %f\n", txConfig->AntennaGain );
				DPRINTF("Channel : %d\n", txConfig->Channel );
				DPRINTF("DataRate : %d\n", txConfig->Datarate );
				DPRINTF("MaxEirp : %f\n", txConfig->MaxEirp );
				DPRINTF("pktLen : %d\n", txConfig->PktLen );
				DPRINTF("TxPower : %d\n", txConfig->TxPower );
			
				for(uint8_t i=0; i<8; i++)
				{
						DPRINTF("TxFreq[%d] : %d\n", i, Channels[i].Frequency );
				}
		}
		else
		{
				
		}

    // Setup the radio frequency
		
    Radio.SetChannel( Channels[txConfig->Channel].Frequency - FREQUENCY_OFFSET );

    Radio.SetTxConfig( MODEM_LORA, phyTxPower, 0, bandwidth, phyDr, 1, 8, false, true, 0, 0, false, 3000 );

    // Setup maximum payload lenght of the radio driver
    Radio.SetMaxPayloadLength( MODEM_LORA, txConfig->PktLen );
    // Get the time-on-air of the next tx frame
    *txTimeOnAir = Radio.TimeOnAir( MODEM_LORA, txConfig->PktLen );

    *txPower = txPowerLimited;
    return true;
}
```

Problem
------------------------------
- log
```
[error]TxFreq : 3
[error]RxFreq : 0
[error]Band : 0
[error]DrRangeMin : -4
[error]DrRangeMax : 2
AntennaGain : 2.150000
Channel : 5
DataRate : 2
MaxEirp : 14.000000
pktLen : 33
TxPower : 7
TxFreq[0] : 2
TxFreq[1] : 44
TxFreq[2] : 2032388
TxFreq[3] : 0
TxFreq[4] : 8
TxFreq[5] : 3
TxFreq[6] : 2032388
TxFreq[7] : 1073876992
```
The error doesn't happen right away.
An error occurs when the device has been operating for a long time
Value breaks are continuous.
What is the cause of this problem?

**Comment by mluis1** [comment]
The version of this project that you are using is now 4 years old.
As a first step I would recommend that you update your project to the latest available version.
Issues have been corrected and improvements have been also done since 4 years.

Up until now we never observed similar issue with this project provided examples. In addition all regions are implemented in a similar way. It has to be noted that there are many end-devices using this stack (on different regions) running for several years without this kind of problems.

From the description it looks like that your implementation is corrupting the RAM memory. 
Most certainly the values are being overwritten by something else in your project.

Please review all the changes you have done in your project in order to try finding what is the root cause for this.

An additional comment:
From the provided code excerpt I noticed that you have modified the KR920 region implementation.
The line:
`Radio.SetChannel( Channels[txConfig->Channel].Frequency - FREQUENCY_OFFSET );`
is different from the original:
https://github.com/Lora-net/LoRaMac-node/blob/f42be67be402a40b3586724800771bfe13fb18e6/src/mac/region/RegionKR920.c#L601

The files under `src/mac` directory should not be modified (furthermore there is no need to it). The provided files have been trough the LoRa-Alliance certification process which means that if there are changes one must verify again that everything is still compliant.

In the future it would be nice if you could post this kind of questions on the project [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab. It is a better place to engage discussions and then we can agree if it is an issue or not.
If you don't mind I move this issue to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.
